Reduce code duplication in OperatorEntry by keying hash map on optional<DispatchKey>

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34160 Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema.
* #33838 Beef up documentation on Dispatcher.h, reorder methods for clarity.
* **#33817 Reduce code duplication in OperatorEntry by keying hash map on optional<DispatchKey>**

Then, nullopt denotes catch all, whereas everything else is specific to
a DispatchKey.  I can delete the second copy of methods when I do this.
This refactor should be pushed all the way to the frontend but I am doing
it one step at a time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20125163](https://our.internmc.facebook.com/intern/diff/D20125163)

ezyang committed:
Reduce code duplication in OperatorEntry by keying hash map on optional<DispatchKey>

Then, nullopt denotes catch all, whereas everything else is specific to
a DispatchKey.  I can delete the second copy of methods when I do this.
This refactor should be pushed all the way to the frontend but I am doing
it one step at a time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -146,12 +146,12 @@ void Dispatcher::deregisterBackendFallbackKernel_(DispatchKey dispatchKey) {
 RegistrationHandleRAII Dispatcher::registerKernel(const OperatorHandle& op, DispatchKey dispatch_key, KernelFunction kernel) {
   // note: this doesn't need the mutex to protect the iterator because write operations on the list keep iterators intact.
-  return op.operatorIterator_->op.registerKernel(std::move(dispatch_key), std::move(kernel));
+  return op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel));
```
No longer need the move bc now implicitly converting to optional? or never needed the move?

smessmer replied:
We never needed the move, dispatch keys are small

bhosmer reviewed (approved):
Ooo nice simplification! 

I can't see how pushing the catchall/non-catchall distinction down costs us anything important - can't imagine the fatter keys will ever make a noticeable difference, and it's reversible if we ever need to hoist the distinction back out to the top level for some reason.

One question about a `std::move` removal inline and definitely worth @smessmer putting the magnifying glass to whatever seething nest of implicit conversions I'm probably glossing over

smessmer reviewed (changes requested):
Not sure this is a move in the right direction. Shouldn't we start using CompoundOpKey and use that to get rid of catchall kernels entirely instead of merging the catchall kernel API into the regular kernel API?

ezyang commented:
> Not sure this is a move in the right direction. Shouldn't we start using CompoundOpKey and use that to get rid of catchall kernels entirely instead of merging the catchall kernel API into the regular kernel API?

Two reasons:
1. The highest priority refactor I have in mind right now is enabling schema-less impl-calls (i.e., this is on the schema merge path). I don't really care about this change per se, but it is very difficult to work on the dispatcher because a lot of codepaths are duplicated. This removes some of the duplication, making further refactors easier to do.
2.  There's a yak in the way of CompoundOp, which is coming up with a plan for kernels that use "catch all" in conjunction with Variable. This is something we should do, but it doesn't seem like it is required for the highest priority refactor (described above). I can't fix everything, so I  have to pick what I'm going to kill first. I choose not to kill catchall.

bhosmer commented:
> Not sure this is a move in the right direction. Shouldn't we start using CompoundOpKey and use that to get rid of catchall kernels entirely instead of merging the catchall kernel API into the regular kernel API?

If IIUC and CompoundOpKey is a kind of DispatchKey - so the new API would be unitary in the same way as it is here, just trading CompoundOpKey for `nullopt` - then I'm not sure I see why this isn't a move in the right direction...

smessmer commented:
Ok I guess I'm fine with this. We'd have to revert the API changes introduced here once we start using CompoundOpKey, but that's not much effort.

bhosmer commented:
> Ok I guess I'm fine with this. We'd have to revert the API changes introduced here once we start using CompoundOpKey, but that's not much effort.

Just to make sure I'm not missing anything though, we'd only have to revert the change to `registerKernal` to take an optional, right? We wouldn't have to reintroduce the parallel fallback API. (I.e., that work would amortize in the move to the final state)

smessmer commented:
> > Ok I guess I'm fine with this. We'd have to revert the API changes introduced here once we start using CompoundOpKey, but that's not much effort.
> 
> Just to make sure I'm not missing anything though, we'd only have to revert the change to `registerKernal` to take an optional, right? We wouldn't have to reintroduce the parallel catchall API. (I.e., that work would amortize in the move to the final state)

yes that's correct

ezyang committed:
Update on "Reduce code duplication in OperatorEntry by keying hash map on optional<DispatchKey>"


Then, nullopt denotes catch all, whereas everything else is specific to
a DispatchKey.  I can delete the second copy of methods when I do this.
This refactor should be pushed all the way to the frontend but I am doing
it one step at a time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20125163](https://our.internmc.facebook.com/intern/diff/D20125163)

[ghstack-poisoned]

ezyang commented:
I updated this to plumb the change all the way through, turns out it was only a few more lines.

ezyang committed:
Update on "Reduce code duplication in OperatorEntry by keying hash map on optional<DispatchKey>"


Then, nullopt denotes catch all, whereas everything else is specific to
a DispatchKey.  I can delete the second copy of methods when I do this.
This refactor should be pushed all the way to the frontend but I am doing
it one step at a time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20125163](https://our.internmc.facebook.com/intern/diff/D20125163)

[ghstack-poisoned]
